scarlet/transformations.py

```diff
@@ -15,147 +16,437 @@ def check_cache(name, key):
+    for n, weight in enumerate(weights):
+        result[slices[n]] += weight * X[inv_slices[n]]
+    return result
+
+class LinearFilter:
+    """A filter that can be applied to an image
+
+    This acts like a sparse diagonal matrix that applies an
+    image of weights to a 2D matrix.
+    """
+    def __init__(self, img, coords):
+        """Initialize the Filter
 
         Parameters
         ----------
-        psf: array-like, default=`None`
-            PSF image in either a single band (used for all images)
-            or an array/list of images with a PSF image for each band.
+        img: 2D or 1D array-like
+            Weights to apply to the filter
+        coords: 2D or 1D array-like
+            Relative coordinates from the current pixel
+            for each weight in `img`
+            (so [0,0] is the current pixel).
         """
-        self.psf = psf
-        self.cache = {}
-
-    def _make_matrices(self, shape, offset_int):
-        """Build Tx, Ty, P, Gamma
+        self.img = np.array(img)
+        self._coords = np.array(coords)
+        self.slices, self.inv_slices = get_filter_slices(self._coords.reshape(-1, 2))
+    
+    @property
+    def T(self):
+        """Transpose the filter
+        """
+        return LinearFilter(self.img, -self._coords)
 
-        To save processing time we separate the diagonal component
-        (which is just an identity matrix) and the off-diagonal
-        component (which is just a band diagonal matrix that is all zeros).
+    def dot(self, X):
+        """Apply the filter to an image or combine filters
 
         Parameters
         ----------
-        shape: tuple
-            Shape of the `~scarlet.Source` frame
-        offset_int: int, default=`None`
-            Integer offset of the translation. See `__init__`.
-
+        X: 2D numpy array or `LinearFilter` or `LinearFilterChain`
+            Array to apply the filter to, or chain of filters to
+            prepend this filter to.
+        
         Returns
         -------
-        result: tuple
-            tx, tx_plus, tx_minus, ty, ty_plus, ty_minus, P
-        tx, ty: `~scipy.sparse` array
-            Sparse arrays that contain the diagonal components of the Tx, Ty matrices
-        tx_plus, ty_plus: `~scipy.sparse` array
-            Sparse arrays that contain the off diagonal components of the Tx, Ty matrices.
+        result: 2D numpy array or `LinearFilterChain`
+            If `X` is an array, this is the result of the
+            filter applied to `X`.
+            If `X` is not an image but is another filter
+            (or chain of filters) then a new `LinearFilterChain` is
+            returned with this one prepended.
+        """
+        if isinstance(X, LinearFilter):
+            return LinearFilterChain([self, X])
+        elif isinstance(X, LinearFilterChain):
+            X.filters.insert(0,self)
+            return X
+        else:
+            return self.apply_filter(X)
```

Comment:
Let this actually be the call to the module-level `apply_filter`, which may or may not be implemented in C++.